### PR TITLE
[fix] pylint errors on module packaging os

### PR DESCRIPTION
**Pull request by 4383**
##### SUMMARY
fix pylint errors

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/packaging/os/*.py

##### ANSIBLE VERSION

```
ansible 2.5.0 (pylint/packaging_os efb8d84c63) last updated 2017/09/22 17:39:58 (GMT +200)
  config file = None
  configured module search path = ['/home/herve/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/herve/dev/ansible/lib/ansible
  executable location = /home/herve/dev/ansible/bin/ansible
  python version = 3.5.2 (default, Nov 17 2016, 17:05:23) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

before changes
```
$ ansible-test sanity --python 3.5 --test pylint lib/ansible/modules/packaging/os/
Sanity check using pylint
ERROR: Found 34 pylint issue(s) which need to be resolved:
ERROR: lib/ansible/modules/packaging/os/apk.py:171:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/apk.py:197:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/apt_repository.py:192:8: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/apt_repository.py:473:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/apt_rpm.py:88:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/apt_rpm.py:134:7: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/homebrew.py:260:8: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/homebrew_cask.py:212:8: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/pacman.py:323:29: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/pkg5_publisher.py:198:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/pkgin.py:191:12: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/pkgin.py:212:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/pkgin.py:235:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/pkgin.py:291:8: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/pkgutil.py:77:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/redhat_subscription.py:308:12: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/redhat_subscription.py:355:8: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/redhat_subscription.py:578:8: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/redhat_subscription.py:615:15: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/rpm_key.py:146:8: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/slackpkg.py:81:7: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/svr4pkg.py:115:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/swdepot.py:117:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/swdepot.py:127:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/urpmi.py:105:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/urpmi.py:156:7: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/urpmi.py:185:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/xbps.py:128:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/xbps.py:141:11: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/xbps.py:171:7: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/xbps.py:192:7: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/yum.py:675:4: no-else-return Unnecessary "else" after "return"
ERROR: lib/ansible/modules/packaging/os/yum_repository.py:588:11: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: lib/ansible/modules/packaging/os/zypper_repository.py:215:7: len-as-condition Do not use `len(SEQUENCE)` as condition value
ERROR: The 1 sanity test(s) listed below (out of 1) failed. See error output above for details.
pylint
```

after changes
```
Sanity check using no-wildcard-import
Sanity check using pep8
Sanity check using pylint
Sanity check using pylint-ansible-test
```

### CODE REVIEW

**Comment by evgkrsk** [comment]
shitpit

**Comment by 4383** [comment]
@KimNorgaard @L2G @alikins @andrew-d @berenddeschouwer @brontitall @danieljaouen @dermute @elasticdog @enriclluelles @hacosta @indrajitr @jasperla @jtyr @kahowell @kbrebanov @matze @mavit @melodous @mkrizek @pmakowski @robinro @sashka @scathatheworm @szinck @tdtrask @troy2914 @verm666 Hello you are  contributors on this module can you review and add shipit if it's look good for you please.

**Comment by kahowell** [comment]
shipit

File: lib/ansible/modules/packaging/os/zypper_repository.py
Comment (robinro) [inline lib/ansible/modules/packaging/os/zypper_repository.py]: I find the old style more readable in the context of the following `elif` statements.

**Review by robinro — CHANGES_REQUESTED** [review]

**Comment by 4383** [comment]
@robinro code review applied!

**Review by robinro — APPROVED** [review]

**Comment by robinro** [comment]
shipit

**Comment by 4383** [comment]
@evgkrsk can you resend your shipit please ? you have tip "shitpit"

**Comment by evgkrsk** [comment]
shipit

**Comment by 4383** [comment]
@bcoca @abadger can you merge this PR please

File: lib/ansible/modules/packaging/os/apk.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/apk.py]: This one should be return result

File: lib/ansible/modules/packaging/os/apt_repository.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/apt_repository.py:465]: I would much prefer we return None explicitly rather than use the implicit None of falling off the end of the function.

On the other hand, now that we only support python-2.6 and above we should just use partial here.

``` python
from functools import partial
[...]
def get_add_ppa_signing_key_callback(module):
    if module.check_mode:
        return None
    return partial(module._run_command(command, check_rc=True)
```

File: lib/ansible/modules/packaging/os/homebrew.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/homebrew.py]: ```valid_state()``` appears to be unused so just delete it.  (The valid states that it lists are out of sync with the argspec so it would cause bugs if it was to be used.)

File: lib/ansible/modules/packaging/os/homebrew_cask.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/homebrew_cask.py]: ```valid_state()``` appears to be unused so delete it.

File: lib/ansible/modules/packaging/os/pkg5_publisher.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/pkg5_publisher.py]: I like to leave ```elif``` when it is logical.  You might as well change the logic of the initial condition to match.  I like tuples.  All of that yields:
``` python
if val in ("-", ""):
    return None
elif val in ("true", "false"):
    return val == "true"
return val
```

File: lib/ansible/modules/packaging/os/rpm_key.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/rpm_key.py:142]: The previous line sets ret to an uppercase version of the string so we only need to check with the uppercase version in this conditional:
``` python
ret = keyid.strip().upper()
if ret.startswith('0X'):
    return ret[2:]
return ret
```

**Comment by bcoca** [comment]
-1 most of these are not errors but style changes, we currently only require PEP8 compiance, not pylint compliance.

File: lib/ansible/modules/packaging/os/xbps.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/xbps.py]: Shouldn't this be ```if not stdout``` ?

**Review by abadger — CHANGES_REQUESTED** [review]

**Comment by 4383** [comment]
@abadger comments review applied! Please check

**Review by abadger — APPROVED** [review]

**Comment by abadger** [comment]
All of my requests are taken care of (except switching to use of partial but that could be done in a separate, more targeted PR.)  I'm kicking this over to @mattclay, @gundalow and the testing working group as to whether we want this.

I think that ansible-test is generating these pylint errors for you because you're using a newer version of pylint than we are (./test/runner/requirements/constraints.txt   lists pylint >= 1.5.3, < 1.7.0 # 1.4.1 adds JSON output, but 1.5.3 fixes bugs related to JSON output  ).

My own evaluation of the classes of changes here:

* I do like len-as-condition as it is a small optimization in addition to being better looking python.  Since we're not doing this in any hot loops it remains a micro-optimization rather than anything earth shattering.
* Fixing no-else-return is definitely more idiomatic python but as bcoca says, this one is purely style.
* Removing dead code (which pylint didn't complain about here but was something that we changed) is "style" but something that we definitely should do.  Dead code carries a cost in that people end up examining it, thinking about what it does, and sometimes modifying it to get rid of bad code habits even though it doesn't do anything.

**Comment by 4383** [comment]
@abadger Thanks for your analysis.
Effectively my version of pylint is greater than your pylint version.
```
$ pylint --version
pylint 1.7.2, 
astroid 1.5.3
Python 3.5.2 (default, Nov 17 2016, 17:05:23) 
[GCC 5.4.0 20160609]
```
We can start to prepare a smooth update of pylint by introducing some tests with the newer version.

**Comment by mattclay** [comment]
As much as I like to see improved code readability and style, we should limit style changes to those enforced by CI. We may upgrade to a newer pylint version in the future (once issues with pylint hanging are resolved). However, that doesn't mean we'll enable all the new rules available in the newer version.

I don't consider removing dead code to be a style change, so those changes appear to be OK. However, I would feel better about those changes if we had at least basic integration test coverage for the affected modules.

**Comment by mattclay** [comment]
We are now using pylint 1.7.4 in CI. All enabled rules are passing.

**Comment by 4383** [comment]
@mattclay how I can help you on my side?

**Comment by mattclay** [comment]
@4383 I've added this PR to the meeting agenda for the Ansible Core Team meeting since the relevant pylint rules are style related. If the consensus is to enable the rules, then these changes will be accepted.

**Comment by mattclay** [comment]
@4383 If you're interested in helping out with pylint related fixes in general, I recommend you take a look at the list of rules that are disabled to see if you want to try working on fixes for any of the non-style related rules. For any significant fixes there you'll probably want to discuss in the Testing Working Group before getting too far into things.

**Comment by dagwieers** [comment]
These files are not PEP8 compliant, and I think both are already tested by the PEP8 sanity tests.

**Comment by mattclay** [comment]
@dagwieers The PR passed the pycodestyle test in CI. What PEP 8 issue(s) are you referring to?

**Comment by dagwieers** [comment]
@mattclay I was probably using pylint with all tests enabled before.

**Comment by mattclay** [comment]
Enabling the pylint rules was discussed in the Ansible Core Team meeting today. The general consensus was in favor of enabling the rules, but we decided to wait until the meeting next week to make a final decision when more people are around to participate in the vote.

**Comment by mattclay** [comment]
@4383 We've agreed to enable the `no-else-return` and `len-as-condition` pylint rules.

Please update the pyllint configuration as part of your PR: https://github.com/ansible/ansible/blob/devel/test/sanity/pylint/config/default

**Comment by 4383** [comment]
@mattclay pylint rules applied!
CI exit with failures due to new pylint rules enabled. Others modules needs refactor for fix theme. I can help you and workon when this PR will be merged.

**Comment by mattclay** [comment]
The scope of changes required (1,174 lines across 479 files) is significant. It may be larger than most people expected when enabling the rules was discussed.

It is not practical to submit or review such large changes in a single PR. Also, it's probably not time well spent, considering there are likely more important pylint rules that could be worked on.

@4383 If you still want to proceed with enabling these two rules, I suggest that we approach it differently. Instead of making all the changes necessary to enable the rules, wait for ansible-test to support excluding rules on a file-by-file basis. With that feature in place, the rules could be enabled with all files that do not pass added to the exclusion list. This would enforce the rules for new files, as well as existing files not on the exclusion list. As time goes by and other files are updated, they could be removed from the exclusion list.

This is how we've handled our PEP 8 cleanup process, which has worked quite well. It allows us to raise the standard for new code without having to make the often significant changes required to bring existing code up to the new standard.

**Comment by 4383** [comment]
@mattclay I like your suggested approach. I wait for ansible-test to support excluding rules on a file-by-file basis.

**Comment by 4383** [comment]
@mattclay How I can follow the ansible-test support progress (irc-meeting, etc...)?

**Comment by gundalow** [comment]
@4383 https://github.com/ansible/community/labels/testing

**Comment by mattclay** [comment]
@4383 I've created an issue to track the per-file exclusions feature: https://github.com/ansible/ansible/issues/34431

**Comment by mattclay** [comment]
@4383 We now have support for ignoring pylint errors on a per file+rule basis.

**Comment by 4383** [comment]
@mattclay Yes I know this is just a dirty test I need to refactor this file

**Comment by maxamillion** [comment]
@4383 are you still interested in contributing this pull request?

**Comment by 4383** [comment]
Hello @maxamillion! Yes

**Comment by maxamillion** [comment]
@4383 can you please rebase on `devel` and resolve the merge conflicts?

**Comment by 4383** [comment]
@maxamillion Yes I try to do this as soon as possible.

**Comment by maxamillion** [comment]
Closing PR due to inactivity, please feel free to re-open and ping me if you desire to continue working on it.

**Pull request closed without merging**
